File: 10192 Vacation/solution.c
#include <stdio.h>
#include <string.h>

int main() {
    int cases = 0;
    char father[102], mother[102];
    while(fgets(father, sizeof(father), stdin) && father[0] != '#') {
        fgets(mother, sizeof(mother), stdin);
        int m = strlen(father) - 1, n = strlen(mother) - 1;
        int L[n + 1];
        memset(L, 0, sizeof(L));
        for(int i = 0; i < m; ++i) {
            int prev = 0;
            for(int j = 1; j <= n; ++j) {
                int temp = L[j];
                if(father[i] == mother[j - 1])
                    L[j] = prev + 1;
                else
                    L[j] = L[j] > L[j - 1] ? L[j] : L[j - 1];
                prev = temp;
            }
        }
        printf("Case #%d: you can visit at most %d cities.\n", ++cases, L[n]);
    }
	return 0;
}

File: 10192 Vacation/solution.java
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int cases = 0;
        String father;
        StringBuilder output = new StringBuilder();
        while((father = br.readLine()) != null && father.charAt(0) != '#') {
            String mother = br.readLine();
            int[] L = new int[mother.length() + 1];
            for(int i = 0; i < father.length(); ++i) {
                int prev = 0;
                for(int j = 1; j <= mother.length(); ++j) {
                    int temp = L[j];
                    if(father.charAt(i) == mother.charAt(j - 1))
                        L[j] = prev + 1;
                    else
                        L[j] = Math.max(L[j], L[j - 1]);
                    prev = temp;
                }
            }
            output.append("Case #").append(++cases).append(": you can visit at most ").append(L[mother.length()]).append(" cities.\n");
        }
	    System.out.print(output);
	}
}
